File: framework/languages/node/e2e/ToActorMessaging/Server/Caller/main.ts
import fs from 'node:fs';
import path from 'node:path';
import { Module } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { ZLinkFrameworkErrorKind, ZLinkFrameworkException, ZLinkMessageFlowLogMode, type ActorRef, type ZLinkActorClient } from '@zlink-systems/framework';
import { ZLINK_ACTOR_CLIENT, ZLinkModule, zlinkFramework } from '@zlink-systems/nestjs';
import { createRedisLocationStore, locationMessagingOptions } from '../../Shared/location-store';
import { actorAsk, actorNotify, actorPush, type ActorCallRequest, type ActorCallResponse, type ActorRefSnapshot, type ActorReply } from '../../Shared/messages';
import { closeHttpServer, startHttpServer } from '../Support/http-server';
import { TO_ACTOR_OPTIONS, createToActorConfigurationModule } from '../../configuration';
import type { ServerOptions } from '../../configuration';

let options: ServerOptions;
let stopping = false;

class CallerModule {}
const configuration = createToActorConfigurationModule();
Module({
  imports: [
    configuration,
    ZLinkModule.forRootFactory({
      imports: [configuration],
      inject: [TO_ACTOR_OPTIONS],
      useFactory: (value: unknown) => {
        options = value as ServerOptions;
        fs.mkdirSync(options.logDir, { recursive: true });
        const builder = zlinkFramework();
        builder.addLocationStore(createRedisLocationStore({
          redisEndpoint: options.redisEndpoint,
          redisKeyPrefix: options.redisKeyPrefix
        }));
        Object.assign(builder.configureLocations(), locationMessagingOptions());
        builder
          .configureDispatch()
          .messageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
          .traceLogFile(path.join(options.logDir, 'caller-flow.log'))
          .traceLabel(options.rid);
        builder
          .addSpotMesh('to-actor')
          .enableRouter(options.routerEndpoint, options.rid)
          .configureEntrySpot({ routingId: options.rid })
          .enablePubSub(options.pubSubEndpoint, options.rid);
        return builder.build();
      }
    })
  ]
})(CallerModule);

async function main(): Promise<void> {
  const app = await NestFactory.createApplicationContext(CallerModule, { logger: false, abortOnError: false });
  const actors = app.get(ZLINK_ACTOR_CLIENT, { strict: false }) as ZLinkActorClient;
  const server = await startHttpServer(options.httpUrl, [
    { method: 'GET', path: '/health', handle: () => ({ status: 'ok' }) },
    {
      method: 'POST',
      path: '/send',
      handle: async (body) => {
        const request = body as ActorCallRequest;
        try {
          await actors
            .sendToActor(requireActorRef(request), actorNotify(request.scenario, request.actorId, request.value))
            .submit();
          return { scenario: request.scenario, actorId: request.actorId, result: 'sent' } satisfies ActorCallResponse;
        } catch (error) {
          return failure(request, error);
        }
      }
    },
    {
      method: 'POST',
      path: '/request',
      handle: async (body) => {
        const request = body as ActorCallRequest;
        try {
          const reply = await actors
            .requestToActor(requireActorRef(request), actorAsk(request.scenario, request.actorId, request.value))
            .timeout(5000)
            .submit<ActorReply>();
          return { scenario: request.scenario, actorId: request.actorId, result: reply.value } satisfies ActorCallResponse;
        } catch (error) {
          return failure(request, error);
        }
      }
    },
    {
      method: 'POST',
      path: '/push',
      handle: async (body) => {
        const request = body as ActorCallRequest;
        try {
          const reply = await actors
            .requestToActor(requireActorRef(request), actorPush(request.scenario, request.actorId, request.value))
            .timeout(5000)
            .submit<ActorReply>();
          return { scenario: request.scenario, actorId: request.actorId, result: reply.value } satisfies ActorCallResponse;
        } catch (error) {
          return failure(request, error);
        }
      }
    },
    { method: 'POST', path: '/shutdown', handle: () => { stopping = true; return { status: 'stopping' }; } }
  ]);

  while (!stopping) {
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  await closeHttpServer(server);
  await app.close();
}

function requireActorRef(request: ActorCallRequest): ActorRef {
  if (request.actor === undefined) {
    throw new ZLinkFrameworkException(
      ZLinkFrameworkErrorKind.ActorRouteNotFound,
      `Actor route '${request.actorId}' was not found.`
    );
  }
  return actorRefFromSnapshot(request.actor);
}

function actorRefFromSnapshot(actor: ActorRefSnapshot): ActorRef {
  return {
    nodeRid: actor.nodeRid,
    actorId: actor.actorId,
    generation: BigInt(actor.generation)
  } as ActorRef;
}

function failure(request: ActorCallRequest, error: unknown): ActorCallResponse {
  return {
    scenario: request.scenario,
    actorId: request.actorId,
    result: 'failed',
    errorKind: error instanceof ZLinkFrameworkException ? error.kind : error instanceof Error ? error.name : String(error)
  };
}

main().catch((error: unknown) => {
  console.error(error);
  process.exitCode = 1;
});

File: framework/languages/cpp/e2e/ToActorMessaging/Server/Caller/main.cpp
/* SPDX-License-Identifier: FSL-1.1-ALv2 */

#include "../../Shared/messages.hpp"
#include "../../Shared/configuration.hpp"

#include <zlink/framework.hpp>
#include <zlink/locations/redis.hpp>

#include <chrono>
#include <memory>
#include <stdexcept>
#include <string>

namespace e2e = zlink::e2e::to_actor_messaging;

namespace
{

void add_redis_location_store (zlink::framework::zlink_framework_options_t &framework,
                               const e2e::redis_configuration_t &redis)
{
    framework.add_location_store (
      std::make_shared<zlink::framework::locations::redis::redis_location_store_t> (
        zlink::framework::locations::redis::redis_location_options_t{
          .connection_string = redis.endpoint, .key_prefix = redis.key_prefix}));
    auto &locations = framework.configure_locations ();
    locations.heartbeat_interval = std::chrono::seconds (1);
    locations.owner_lease_ttl = std::chrono::seconds (3);
    locations.polling_interval = std::chrono::milliseconds (500);
}

std::shared_ptr<zlink::framework::location_store_t>
make_redis_location_store (const e2e::redis_configuration_t &redis)
{
    return std::make_shared<zlink::framework::locations::redis::redis_location_store_t> (
      zlink::framework::locations::redis::redis_location_options_t{
        .connection_string = redis.endpoint, .key_prefix = redis.key_prefix});
}

void write_fault_actor_row (const e2e::actor_call_request_t &request,
                            const e2e::caller_configuration_t &configuration)
{
    auto store = make_redis_location_store (configuration.redis);
    const auto owner_id = "toactor-fault-" + request.actor_id;
    const auto actor_node =
      request.scenario == "TA-B3-prepare" ? std::string ("ghost-node")
                                          : configuration.actor_rid;
    const auto actor_spot =
      request.scenario == "TA-B3-prepare" ? std::string ("ghost-spot")
                                          : configuration.actor_rid;
    const auto lease =
      store->renew_owner_lease (owner_id, zlink::routing_id_t::from (actor_node),
                                std::chrono::seconds (30))
        .result ();
    if (!lease) {
        throw lease.error () ? *lease.error ()
                             : zlink::framework::framework_exception_t (
                                 zlink::framework::framework_error_kind_t::request_failed,
                                 "fault owner lease failed");
    }
    auto write =
      store
        ->update_actor (
          zlink::framework::actor_location_t{
            .actor_id = request.actor_id,
            .actor_type = e2e::actor_type_name,
            .actor_ref = zlink::framework::actor_ref_t (
              zlink::framework::node_rid_t::from_string (actor_node), e2e::actor_type_name,
              request.actor_id, 99),
            .node_rid = zlink::routing_id_t::from (actor_node),
            .location_kind = zlink::spot_kind::entry,
            .spot_mesh_name = e2e::spot_mesh_name,
            .spot_rid = zlink::routing_id_t::from (actor_spot),
            .owner_id = owner_id},
          zlink::framework::location_write_intent_t::takeover)
        .result ();
    if (!write
        || write.value ().status != zlink::framework::location_write_status_t::stored) {
        throw write.error () ? *write.error ()
                             : zlink::framework::framework_exception_t (
                                 zlink::framework::framework_error_kind_t::request_failed,
                                 "fault actor row write failed");
    }
}

std::string kind_name (zlink::framework::framework_error_kind_t kind)
{
    switch (kind) {
        case zlink::framework::framework_error_kind_t::actor_route_not_found:
            return "actor_route_not_found";
        case zlink::framework::framework_error_kind_t::actor_location_stale:
            return "actor_location_stale";
        case zlink::framework::framework_error_kind_t::route_not_connected:
            return "route_not_connected";
        case zlink::framework::framework_error_kind_t::request_failed:
            return "request_failed";
        default:
            return "framework_error_" + std::to_string (static_cast<int> (kind));
    }
}

e2e::actor_call_response_t failed (const e2e::actor_call_request_t &request,
                                   const zlink::framework::framework_exception_t &error)
{
    return {request.scenario, request.actor_id, error.what (), kind_name (error.kind ())};
}

class send_handler_t
{
  public:
    using dependency_types = zlink::framework::dependency_list_t<zlink::framework::actor_client_t,
                                                                zlink::framework::actor_directory_t>;
    using request_type = e2e::actor_call_request_t;
    using reply_type = e2e::actor_call_response_t;

    send_handler_t (zlink::framework::actor_client_t &actors,
                    zlink::framework::actor_directory_t &directory) :
        _actors (actors), _directory (directory)
    {
    }

    zlink::framework::task_t<e2e::actor_call_response_t>
    handle (const e2e::actor_call_request_t &request)
    {
        try {
            e2e::actor_notify_t notify{request.scenario, request.actor_id, request.value};
            auto actor_ref = co_await _directory.find (request.actor_id);
            if (!actor_ref) {
                throw zlink::framework::framework_exception_t (
                  zlink::framework::framework_error_kind_t::actor_route_not_found,
                  "actor route was not found");
            }
            _actors.send_to_actor (*actor_ref, notify)
              .submit ();
            co_return e2e::actor_call_response_t{request.scenario, request.actor_id, "sent", ""};
        }
        catch (const zlink::framework::framework_exception_t &error) {
            co_return failed (request, error);
        }
    }

  private:
    zlink::framework::actor_client_t &_actors;
    zlink::framework::actor_directory_t &_directory;
};

class request_handler_t
{
  public:
    using dependency_types = zlink::framework::dependency_list_t<zlink::framework::actor_client_t,
                                                                zlink::framework::actor_directory_t>;
    using request_type = e2e::actor_call_request_t;
    using reply_type = e2e::actor_call_response_t;

    request_handler_t (zlink::framework::actor_client_t &actors,
                       zlink::framework::actor_directory_t &directory) :
        _actors (actors), _directory (directory)
    {
    }

    zlink::framework::task_t<e2e::actor_call_response_t>
    handle (const e2e::actor_call_request_t &request)
    {
        try {
            e2e::actor_ask_t ask{request.scenario, request.actor_id, request.value};
            auto actor_ref = co_await _directory.find (request.actor_id);
            if (!actor_ref) {
                throw zlink::framework::framework_exception_t (
                  zlink::framework::framework_error_kind_t::actor_route_not_found,
                  "actor route was not found");
            }
            auto reply = co_await _actors.request_to_actor (*actor_ref, ask)
                           .timeout (std::chrono::seconds (5))
                           .async<e2e::actor_reply_t> ();
            co_return e2e::actor_call_response_t{
              request.scenario, request.actor_id, reply.value, ""};
        }
        catch (const zlink::framework::framework_exception_t &error) {
            co_return failed (request, error);
        }
    }

  private:
    zlink::framework::actor_client_t &_actors;
    zlink::framework::actor_directory_t &_directory;
};

class prepare_failure_handler_t
{
  public:
    using dependency_types =
      zlink::framework::dependency_list_t<e2e::caller_configuration_t>;
    using request_type = e2e::actor_call_request_t;
    using reply_type = e2e::actor_call_response_t;

    explicit prepare_failure_handler_t (e2e::caller_configuration_t &configuration) :
        _configuration (configuration)
    {
    }

    e2e::actor_call_response_t handle (const e2e::actor_call_request_t &request)
    {
        try {
            write_fault_actor_row (request, _configuration);
            return {request.scenario, request.actor_id, "prepared", ""};
        }
        catch (const zlink::framework::framework_exception_t &error) {
            return failed (request, error);
        }
    }

  private:
    e2e::caller_configuration_t &_configuration;
};

} // namespace

int main (int argc, char **argv)
{
    auto app = zlink::framework::app_t::create ();
    const auto configuration =
      e2e::load_role_configuration<e2e::caller_configuration_t> (app, argc, argv);
    app.logging ().use_file (configuration.log_dir + "/caller.log").set_min_level (
      zlink::framework::log_level_t::debug);
    app.add_zlink_framework ([configuration] (
                              zlink::framework::zlink_framework_options_t &framework) {
        framework.configure_dispatch ()
          .message_flow (zlink::framework::message_flow_log_mode_t::key_transitions)
          .trace_log_file (configuration.log_dir + "/caller-flow.log")
          .trace_label ("cpp-to-actor-caller");
        framework.services ().add_singleton<e2e::caller_configuration_t> (
          std::make_unique<e2e::caller_configuration_t> (configuration));
        add_redis_location_store (framework, configuration.redis);
        auto mesh = framework.add_spot_mesh (e2e::spot_mesh_name)
          .enable_router (configuration.spot_endpoint)
          .enable_pub_sub (configuration.pub_sub_endpoint)
          .set_routing_id (zlink::routing_id_t::from (configuration.node_rid));
        if (!configuration.actor_spot_endpoint.empty ()) {
            mesh.connect_router (zlink::routing_id_t::from (configuration.actor_rid),
                                 configuration.actor_spot_endpoint);
        }
        framework.http ()
          .listen (configuration.http_endpoint)
          .map_health ("/health")
          .map_post<send_handler_t> ("/send")
          .map_post<request_handler_t> ("/request")
          .map_post<prepare_failure_handler_t> ("/prepare-failure");
    });
    return app.run (argc, argv);
}
